csrc/optimization/consecutive_cast.cpp

```diff
@@ -0,0 +1,194 @@
+//    a straight line dependency:
+//        1.3.1 we point starting_anchor at the beginning op, indicating the
+//        starting point of our folding optimization, meanwhile, we point
+//        lo_anchor at the first op, indicating the narrowest dtype we have seen
+//        in the segment;
+//        1.3.2 we enter the loop to iterate through items
+//        inside `chain_cast_vals`, for item `val`:
+//
+//              a. if `val_dtype` is the same as, or wider than `anchor_dtype`
+//              of `lo_anchor`, current cast is a no-op and can be ignored;
+//
+//              b. if `anchor_dtype` is narrower than `val_dtype`, previous cast
+//              to `lo_anchor` is a no-op and can be folded away. We update
+//              `lo_anchor` to point to `val`;
+//
+//              c. otherwise, `val` and `lo_anchor` are incompatible casts and
+//              both needs to be preserved. We'll rewire it as:
+//              `starting_anchor`->`lo_anchor`->`val`. Afterwards, we'll update
+//              `starting_anchor` and `lo_anchor` to both point at `val`.
+//
+//    1.4 At this point we look at `anchor_dtype` of `lo_anchor` and
+//    `output_dtype` of `expr->output(0)`:
+//
+//        a. if `anchor_dtype` is the same as `output_dtype`, we skip the last
+//        cast op and replace all its uses with `lo_anchor`;
+//
+//        b. if `anchor_dtype` is wider than `output_dtype`, all previous cast
+//        after `starting_anchor` is no-op, we re-wire `starting_anchor`
+//        directly to `expr`;
+//
+//        c. otherwise, we can't bypass `lo_anchor` cast, we rewire this
+//        section as `starting_anchor`->`lo_anchor`->`expr->output(0)`
+void castOptimizationPass(Fusion* fusion) {
+  // TODO: Traveral implies topological order on returned exprs, we can leverage
+  // that to improve the effieciency of the pass. In the case of a straight line
+  // casts, we are doing a lot of meaningless work here on mutating intermediate
+  // casts that would have been done again at the end of the chain.
+  // We should really use the reverse topological order and filters out exprs
+  // that has been rendered as dead code during the pass.
+  for (auto expr : fusion->exprs()) {
+    // skip current expr if it's not a foldable cast
+    if (!isCast(expr)) {
+      continue;
+    }
+    std::list<Val*> chain_cast_vals;
+    auto prev_expr = expr->input(0)->definition();
+    while (isCast(prev_expr)) {
+      auto intermediate_cast = prev_expr->output(0);
+      // 1.2 Note, if the output of prev_expr
+      //   is used by other operation(s); or
+      //   is a direct output from fusion
+      // we skip the casting chaining
+      if (intermediate_cast->isFusionOutput() ||
+          intermediate_cast->uses().size() > 1) {
+        break;
+      }
+
+      // in the loop, we just repetitively chaining consecutive casts.
+      chain_cast_vals.push_front(intermediate_cast);
```

Comment:
Should we mark `intermediate_cast` as visited and skip this expr in outer loop? I think your TODO is pretty easy to implement: just reverse the order of outer loop and use a set marking all visited extra.